### PR TITLE
[POC]: Build treePath in engine (ProcessInstanceRecordValue)

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

> [!Note]
>
> More details [here](https://docs.google.com/document/d/1PrWLli07nLDterUe8AMDAyjs_mICNx0kYQfgC5v3Fro/edit?tab=t.0)

This is one of the POC we planned to do to evaluate what the effort and impact is to build the treePath in different places.

general

In general the performance was looking pretty good, until we encountered an pod restart, which influenced the complete cluster. It caused high CPU throttling and in consequence, all metrics were impacted.

restart-cpu-throttle

When we compare to base, we see that the throughput was for a while quite similar to our base weekly medic benchmarks.
comp-general

This is also for throughput and latency
throughput
latency

For the request sizes, we can see a slight increase (as expected). The increase for P99 was ~10KB and for P90 ~1KB.
requests-sizes

In the CPU panels, we can see similar numbers until encounter our pod restart.

cpu

To validate that the ES is not impacted, we also checked the ES CPU resources and the seem to be comparable.

es-cpu

### Working POC 

In ES we can see that the treePath is correctly set for the list view and FNI index.

list-view-treepath
operate-fni

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

See related POC #25310

### CODE REVIEW

File: zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java [hunk 1]
```diff
@@ -59,6 +66,13 @@ public final class ProcessInstanceRecord extends UnifiedRecordValue
...
 
+  private final ArrayProperty<CompressedLongArrayValue> elementInstancePathProp =
+      new ArrayProperty<>("elementInstancePath", CompressedLongArrayValue::new);
+  private final ArrayProperty<LongValue> processDefinitionPathProp =
+      new ArrayProperty<>("processDefinitionPath", LongValue::new);
+  private final ArrayProperty<IntegerValue> callingElementPathProp =
+      new ArrayProperty<>("callingElementPath", IntegerValue::new);
+
   public ProcessInstanceRecord() {
     super(11);
```
Comment (ChrisKujawa) [inline zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:77]: ❌ Needs to be updated

File: zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java [hunk 2]
```diff
@@ -59,6 +66,13 @@ public final class ProcessInstanceRecord extends UnifiedRecordValue
   private final LongProperty parentElementInstanceKeyProp =
       new LongProperty("parentElementInstanceKey", -1L);
 
+  private final ArrayProperty<CompressedLongArrayValue> elementInstancePathProp =
+      new ArrayProperty<>("elementInstancePath", CompressedLongArrayValue::new);
+  private final ArrayProperty<LongValue> processDefinitionPathProp =
+      new ArrayProperty<>("processDefinitionPath", LongValue::new);
+  private final ArrayProperty<IntegerValue> callingElementPathProp =
+      new ArrayProperty<>("callingElementPath", IntegerValue::new);
```
Comment (ChrisKujawa) [inline zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:74]: 🔧 Add more docs around this

File: zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java [hunk 3]
```diff
@@ -169,6 +186,62 @@ public ProcessInstanceRecord setBpmnEventType(final BpmnEventType bpmnEventType)
...
+  public ProcessInstanceRecord setElementInstancePath(final List<List<Long>> elementInstancePath) {
+    elementInstancePathProp.reset();
+    elementInstancePath.forEach(
+        pathEntry -> {
+          final var entry = elementInstancePathProp.add();
+
+          final long[] longs = new long[pathEntry.size()];
+          for (int i = 0; i < pathEntry.size(); i++) {
+            longs[i] = pathEntry.get(i);
+          }
```
Comment (mustafadagher) [inline zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:211]:
```suggestion
          final long[] longs = pathEntry.stream().mapToLong(Long::longValue).toArray();
```

**Review by mustafadagher — COMMENTED** [review]
I went over the PR and left a few suggestions and questions. It looks good to me overall 🚀

**Comment by ChrisKujawa** [comment]
supersed by https://github.com/camunda/camunda/pull/25501

**Pull request closed without merging**